lib/Optimizer/Transforms/CombineMeasurements.cpp

```diff
@@ -0,0 +1,258 @@
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/CodeGen/QIRAttributeNames.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeTypes.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "nlohmann/json.hpp"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_COMBINEMEASUREMENTS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "combine-measurements"
+
+using namespace mlir;
+
+namespace {
+
+struct Analysis {
+  using OutputNamesType =
+      std::map<std::size_t, std::pair<std::size_t, std::string>>;
+
+  Analysis() = default;
+  Analysis(const Analysis &) = delete;
+  Analysis(Analysis &&) = delete;
+  Analysis &operator=(const Analysis &) = delete;
+
+  mlir::DenseMap<mlir::Value, std::size_t> measurements;
+  OutputNamesType resultQubitVals;
+  quake::MzOp lastMeasurement;
+
+  bool empty() const { return measurements.empty(); }
+
+  LogicalResult analyze(func::FuncOp func) {
+    quake::AllocaOp qalloc;
+    std::size_t currentOffset = 0;
+
+    for (auto &block : func.getRegion()) {
+      for (auto &op : block) {
+        if (auto alloc = dyn_cast_or_null<quake::AllocaOp>(&op)) {
+          if (qalloc)
+            return op.emitError("Multiple qalloc statements found");
+
+          qalloc = alloc;
+        } else if (auto measure = dyn_cast_or_null<quake::MzOp>(&op)) {
+          if (!measure.use_empty()) {
+            measure.emitWarning("Measurements with uses are not supported");
+            return success();
+          }
+
+          auto veqOp = measure.getOperand(0);
+          auto ty = veqOp.getType();
+
+          std::size_t size = 0;
+          if (auto veqTy = dyn_cast<quake::RefType>(ty))
+            size = 1;
+          else if (auto veqTy = dyn_cast<quake::VeqType>(ty)) {
+            size = veqTy.getSize();
+            if (size == 0)
+              return op.emitError("Unknown measurement size");
+          }
+
+          measurements[measure.getMeasOut()] = currentOffset;
+          lastMeasurement = measure;
+          currentOffset += size;
+        }
+      }
+    }
+
+    return success();
+  }
+};
+
+class ExtendQubitMeasurePattern : public OpRewritePattern<quake::MzOp> {
+  using OutputNamesType =
+      std::map<std::size_t, std::pair<std::size_t, std::string>>;
```

Comment:
This looks the same as above. Should it be pulled out separately and used as a common definition?

Reply:
Done, thanks!